Jetpack Search: add option for new inline search

trakos opened the pull request:
## Proposed changes:
* Add site option `swap_classic_to_inline_search`
* Use new Inline Search instead of Classic Search if that option is set to true
* Allow changing that option through `/wp-json/jetpack/v4/search/settings` endpoint
* Automatically switch it on if classic search returns field `swap_classic_to_inline_search` set to true
* Automatically switch if off if instant search returns field `swap_classic_to_inline_search` set to false
* Automatically update it if plan-info returns field `swap_classic_to_inline_search` with boolean value

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pgbpWj-4q-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:

To test the API endpoint I've used an HTTP client from my IDE:

```
GET https://example.com/wp-json/jetpack/v4/search/settings HTTP/1.1
x-wp-nonce: xxx
cookie: wordpress_logged_in_xxx=yyy;
```

```
POST https://example.com/wp-json/jetpack/v4/search/settings HTTP/1.1
Content-Type: application/json
x-wp-nonce: xxx
cookie: wordpress_logged_in_xxx=yyy;

{
  "swap_classic_to_inline_search": true
}
```

You can obtain nonce and cookie by enabling instant search on your site. That should trigger the API request, and you can find those values in the request headers using dev console.

To test API results, you need to use your wpcom sandbox - enable it in wp-config.php:

```
define( 'JETPACK__SANDBOX_DOMAIN', 'your-sandbox-domain' );
```

and apply this patch to modify both API responses and plan-info: 37d30-pb

Now, once you visit `/?s=test` on your test site, it should rotate between v1 and v1.3 API. Remove `revert_to_classic_search` from v1.3 output in order to use new Inline Search.

Additionally, once you visit search dashboard in admin it should switch to new Inline Search and v1.3 API.

trakos committed:
Jetpack Search: add option for new inline search

gibrown commented on projects/packages/search/src/class-module-control.php:
```diff
@@ -33,9 +33,10 @@ class Module_Control {
+	const JETPACK_ACTIVE_MODULES_OPTION_KEY              = 'active_modules';
+	const JETPACK_SEARCH_MODULE_SLUG                     = 'search';
+	const SEARCH_MODULE_INSTANT_SEARCH_OPTION_KEY        = 'instant_search_enabled';
+	const SEARCH_MODULE_USE_NEW_INLINE_SEARCH_OPTION_KEY = 'use_new_inline_search';
```
Hmmm... naming... Should we call this swap_classic_to_inline search or something. I know we'll eventually delete this option so the "new" kinda works, but right now it sounds a bit too much like this and instant search enabled are opposites.

trakos replied:
I've renamed it as suggested.

gibrown commented on projects/packages/search/src/class-module-control.php:
```diff
@@ -76,6 +77,15 @@ public function is_instant_search_enabled() {
+	 * @return bool
+	 */
+	public function is_using_new_inline_search() {
+		return (bool) get_option( self::SEARCH_MODULE_USE_NEW_INLINE_SEARCH_OPTION_KEY );
```
Should this explicitly specify a default (second argument) so if the option doesn't exist then it is false?

trakos replied:
False is [the default value for the $default_value argument](https://github.com/WordPress/WordPress/blob/master/wp-includes/option.php#L78), so it's already the case.

trakos replied:
Added `false` to clarify

gibrown commented on projects/packages/search/src/classic-search/class-classic-search.php:
```diff
@@ -389,6 +389,9 @@ public function search( array $es_args ) {
 		}
 
 		$response = json_decode( wp_remote_retrieve_body( $request ), true );
+		if ( isset( $response['use_new_inline_search'] ) && $response['use_new_inline_search'] === true ) {
```
The part of this I dislike the most is that we'll effectively be doing DB writes from front end traffic. We could leave it in though. I guess it gives us an option.

trakos replied:
It's [ignored when it's already the same value](https://github.com/WordPress/WordPress/blob/master/wp-includes/option.php#L915-L926), so it should trigger only one update per site.  We could do it when someone visits search dashboard instead, but the rollout would be much slower then.

trakos replied:
I've added additional handling in the plan-info request handling, so we will be able to trigger the update when someone visits search dashboard in wp-admin. The rollout would be slower, but the update would only happen from admin traffic.

This gives us another option - we can decide which one to use later.

gibrown replied:
Ya, I think it is good to have this option in there and agree it should be fine because it will only be getting set once.

gibrown reviewed (commented):
Looks correct. Haven't tested. Mostly just had some thoughts on the naming because that is "fun" to argue about. ;)

@jeherve @kraftbj I think it may make sense for one of you (or someone else) to look at this. We're going to use this mechanism to eventually switch from using the classic search api to instant search api for inline search queries. Either we'll write the option remotely from a script on wpcom, or we'll update the search API to trigger the switch.

trakos committed:
Changes after review:

- rename use_new_inline_search to swap_classic_to_inline_search
- allow changing the setting through plan info endpoint

gibrown reviewed (dismissed):
I'm blocked on testing still, but I think all of this looks correct. Shouldn't change any functionality until we enable it. Getting some review from other Jetpack folks would still be good though.

jeherve commented on projects/packages/search/src/class-rest-controller.php:
```diff
@@ -268,6 +269,13 @@ public function update_settings( $request ) {
 
+		if ( $swap_classic_to_inline_search !== null ) {
+			$swap_classic_to_inline_search_updated = $this->search_module->update_swap_classic_to_inline_search( $swap_classic_to_inline_search );
+			if ( is_wp_error( $swap_classic_to_inline_search_updated ) ) {
```
Will we get to that at any point? I wouldn't expect `update_option` to run a `WP_Error`.

trakos replied:
No, you're right, `update_option` won't return an error. The other methods in this file had some custom checks, but we didn't need it here, so I've removed this `is_wp_error` check.

jeherve reviewed (dismissed):
This tested well for me, I think it should be good to go. I have a question below.

********

I should note that at this point, things are getting a bit confusing, between Classic, Inline, and Instant. At least they're getting confusing to me 😅 

I think a few lines about the different options, why they're there, what they do, in the package's readme could be helpful for new contributors.

> To test the API endpoint I've used an HTTP client from my IDE:

I personally found it easier to update the value from my browser console, while in the block editor on the site:

```js
wp.apiFetch( { method: 'POST', path: 'jetpack/v4/search/settings', data: { swap_classic_to_inline_search: true } } ).then( data => console.log( data ) );
```

> Now, once you visit /?s=test on your test site, it should rotate between v1 and v1.3 API. Remove revert_to_classic_search from v1.3 output in order to use new Inline Search.

I wasn't immediately sure what that meant. 

I ended up testing like so:

1. I enabled Search, but disabled instant search
2. I installed the Debug bar plugin
3. I checked the URL of the query that happened when loading a search page on the site.
4. In another window in the post editor, I updated the `swap_classic_to_inline_search` value from false to true, and back, using the code I mentioned above.
5. Every time I switched things, I would then refresh the search page and check the query URL.

trakos committed:
Update readme

trakos committed:
Merge remote-tracking branch 'origin/trunk' into add/inline-search-option

trakos committed:
Remove redundant error check

trakos commented:
Thank you for the review and testing this!

> I think a few lines about the different options, why they're there, what they do, in the package's readme could be helpful for new contributors.

Good point! I've added a short description for each of the search modes. This new Inline search will replace Classic, so hopefully we will be able to reduce it to two options eventually. This new setting is only required for the feature rollout.

gibrown commented:
> I should note that at this point, things are getting a bit confusing, between Classic, Inline, and Instant. At least they're getting confusing to me 😅

Just to clarify here @jeherve the plan is to completely remove Classic once we have switched to inline. At that point the only difference between the two will be one is in a modal and the other is handled in the theme. They'll both use the same API and the backend can support the same features.
